File: clang/sort/heap_sort.c
#include <stdio.h>

void heapify(int arr[], int n, int i) {
  int temp, maximum;
  int left_index, right_index;

  maximum = i;
  right_index = 2 * i + 2;
  left_index = 2 * i + 1;

  if (left_index < n && arr[left_index] > arr[maximum])
    maximum = left_index;

  if (right_index < n && arr[right_index] > arr[maximum])
    maximum = right_index;

  if (maximum != i) {
    temp = arr[i];
    arr[i] = arr[maximum];
    arr[maximum] = temp;
    heapify(arr, n, maximum);
  }
}

void heap_sort(int arr[], int n) {
  int i, temp;
  for (i = n / 2 - 1; i >= 0; i--) {
    heapify(arr, n, i);
  }

  for (i = n - 1; i > 0; i--) {
    temp = arr[0];
    arr[0] = arr[i];
    arr[i] = temp;
    heapify(arr, i, 0);
  }
}

int main() {
  int length = 6;
  int arr[] = {20, 18, 5, 15, 3, 2};

  printf("Original Array : ");
  for (int i = 0; i < length; i++) {
    printf("%d ", arr[i]);
  }

  printf("\n");
  heap_sort(arr, length);

  printf("Array after performing heap sort: ");
  for (int i = 0; i < length; i++) {
    printf("%d ", arr[i]);
  }
  return 0;
}

File: nim/sort/heap_sort.nim
proc heapify(arr: var openArray[int], size: int, i: int) =
  var maximum = i
  let leftIndex = 2 * i + 1
  let rightIndex = 2 * i + 2

  if leftIndex < size and arr[leftIndex] > arr[maximum]:
    maximum = leftIndex

  if rightIndex < size and arr[rightIndex] > arr[maximum]:
    maximum = rightIndex

  if maximum != i:
    swap(arr[i], arr[maximum])
    arr.heapify(size, maximum)

proc heapSort(arr: var openArray[int]) =
  for i in countdown((arr.len div 2) - 1, 0):
    arr.heapify(arr.len, i)

  for i in countdown(arr.len - 1, 0):
    swap(arr[0], arr[i])
    arr.heapify(i, 0)

proc main() =
  var arr = [9,20,7,18,5,15,3,2]

  echo "Original Array: "
  for i in 0 ..< arr.len():
    stdout.write $arr[i] & " "
  echo()

  arr.heapSort()
  
  echo "Array after performing heap sort: "
  for i in 0 ..< arr.len():
    stdout.write $arr[i] & " "
  echo()

when isMainModule:
  main()
